File: Backend/discussion Module/discussion.js
const $ = (id) => document.getElementById(id);

let discussions = [];
let organizations = [];
let selectedDiscussionId = null;
let searchTimer = null;

let currentUser = JSON.parse(localStorage.getItem("user") || "{}");
let currentUserId = Number(currentUser.id || currentUser.user_id);

function safe(value) {
    return String(value || "").replace(/[&<>'"]/g, (character) => ({
        "&": "&amp;",
        "<": "&lt;",
        ">": "&gt;",
        "'": "&#39;",
        '"': "&quot;"
    }[character]));
}

function toast(message) {
    $("toast").textContent = message;
    $("toast").classList.add("show");
    setTimeout(() => $("toast").classList.remove("show"), 2800);
}

function getInitials(name) {
    return String(name || "U")
        .trim()
        .split(/\s+/)
        .slice(0, 2)
        .map((part) => part.charAt(0).toUpperCase())
        .join("") || "U";
}

async function requireLoggedInUser() {
    if (Number.isInteger(currentUserId) && currentUserId > 0) {
        return true;
    }

    try {
        const response = await fetch("http://127.0.0.1:5001/check-login", {
            credentials: "include"
        });
        const data = await response.json();

        if (data.logged_in && data.user?.id) {
            currentUser = data.user;
            currentUserId = Number(data.user.id);
            localStorage.setItem("user", JSON.stringify(data.user));
            return true;
        }
    } catch (error) {
        console.error("Login check failed:", error);
    }

    window.location.href = "http://127.0.0.1:5001/";
    return false;
}

function buildQuery() {
    const params = new URLSearchParams();
    params.set("user_id", currentUserId);

    const search = $("searchInput").value.trim();
    const orgId = $("organizationFilter").value;

    if (search) {
        params.set("search", search);
    }

    if (orgId) {
        params.set("org_id", orgId);
    }

    return params.toString();
}

function renderOrganizationOptions() {
    const filterOptions = ['<option value="">All organizations</option>'];
    const formOptions = ['<option value="">Choose organization</option>'];

    organizations.forEach((organization) => {
        const option = `<option value="${organization.org_id}">${safe(organization.name)}</option>`;
        filterOptions.push(option);
        formOptions.push(option);
    });

    $("organizationFilter").innerHTML = filterOptions.join("");
    $("discussionOrganization").innerHTML = formOptions.join("");
    $("orgCount").textContent = organizations.length;
}

function renderStats() {
    $("totalCount").textContent = discussions.length;
    $("myCount").textContent = discussions.filter((discussion) => discussion.is_owner).length;
}

function renderEmptyDetail() {
    selectedDiscussionId = null;
    $("detailPanel").innerHTML = `
        <div class="empty-detail">
            <i class="fa-solid fa-comments"></i>
            <h2>Select a discussion</h2>
            <p>Open a discussion to read the full post and manage your own entries.</p>
        </div>
    `;
}

function renderDiscussions() {
    renderStats();

    if (!discussions.length) {
        $("discussionList").innerHTML = '<p class="state-message">No discussions found. Start the first conversation for your community.</p>';
        renderEmptyDetail();
        return;
    }

    $("discussionList").innerHTML = discussions.map((discussion) => `
        <button class="discussion-card ${discussion.discussion_id === selectedDiscussionId ? "active" : ""}" type="button" onclick="openDiscussion(${discussion.discussion_id})">
            <div class="discussion-meta">
                <span class="pill"><i class="fa-solid fa-building-user"></i> ${safe(discussion.organization_name)}</span>
                <span><i class="fa-regular fa-calendar"></i> ${safe(discussion.created_at)}</span>
            </div>
            <h3>${safe(discussion.title)}</h3>
            <p>${safe(discussion.content)}</p>
        </button>
    `).join("");
}

function renderDiscussionDetail(discussion) {
    selectedDiscussionId = discussion.discussion_id;
    renderDiscussions();

    $("detailPanel").innerHTML = `
        <div class="detail-header">
            <div class="detail-title-row">
                <div>
                    <span class="eyebrow">${safe(discussion.organization_category)}</span>
                    <h2>${safe(discussion.title)}</h2>
                </div>
                <span class="pill"><i class="fa-solid fa-building-user"></i> ${safe(discussion.organization_name)}</span>
            </div>
            <div class="author-row">
                <span class="author-avatar">${safe(getInitials(discussion.author))}</span>
                <span>Posted by <strong>${safe(discussion.author)}</strong> on ${safe(discussion.created_at)}</span>
            </div>
        </div>
        <div class="detail-content">${safe(discussion.content)}</div>
        <div class="detail-actions">
            ${discussion.is_owner ? `
                <button class="secondary-button" type="button" onclick="openEditPanel(${discussion.discussion_id})">
                    <i class="fa-solid fa-pen"></i> Edit
                </button>
                <button class="danger-button" type="button" onclick="deleteDiscussion(${discussion.discussion_id})">
                    <i class="fa-solid fa-trash"></i> Delete
                </button>
            ` : '<span class="pill"><i class="fa-solid fa-lock"></i> Author only controls</span>'}
        </div>
    `;
}

async function loadOrganizations() {
    try {
        const response = await fetch("/api/organizations");
        const data = await response.json();
        if (!response.ok) {
            throw new Error(data.message || "Unable to load organizations.");
        }

        organizations = data.organizations || [];
        renderOrganizationOptions();
    } catch (error) {
        organizations = [];
        renderOrganizationOptions();
        toast(error.message || "Organizations could not be loaded.");
    }
}

async function loadDiscussions() {
    $("discussionList").innerHTML = '<p class="state-message">Loading discussions...</p>';

    try {
        const response = await fetch(`/api/discussions?${buildQuery()}`);
        const data = await response.json();
        if (!response.ok) {
            throw new Error(data.message || "Unable to load discussions.");
        }

        discussions = data.discussions || [];
        const selected = discussions.find((discussion) => discussion.discussion_id === selectedDiscussionId);
        renderDiscussions();

        if (selected) {
            renderDiscussionDetail(selected);
        } else if (!discussions.length) {
            renderEmptyDetail();
        }
    } catch (error) {
        discussions = [];
        renderStats();
        $("discussionList").innerHTML = '<p class="state-message">Discussions could not be loaded. Please check the Flask server and database.</p>';
        renderEmptyDetail();
        toast(error.message || "Unable to load discussions.");
    }
}

async function openDiscussion(discussionId) {
    try {
        const response = await fetch(`/api/discussions/${discussionId}?user_id=${currentUserId}`);
        const data = await response.json();
        if (!response.ok) {
            throw new Error(data.message || "Unable to open discussion.");
        }

        renderDiscussionDetail(data.discussion);
    } catch (error) {
        toast(error.message || "Unable to open discussion.");
    }
}

function openCreatePanel() {
    $("discussionForm").reset();
    $("discussionId").value = "";
    $("discussionOrganization").disabled = false;
    $("modalEyebrow").textContent = "Start a conversation";
    $("modalTitle").textContent = "Create discussion";
    $("saveButton").textContent = "Create discussion";
    $("discussionModal").classList.add("show");
}

function openEditPanel(discussionId) {
    const discussion = discussions.find((item) => item.discussion_id === discussionId);
    if (!discussion) {
        toast("Please open this discussion again before editing.");
        return;
    }

    $("discussionId").value = discussion.discussion_id;
    $("discussionOrganization").value = discussion.org_id;
    $("discussionOrganization").disabled = true;
    $("discussionTitle").value = discussion.title;
    $("discussionContent").value = discussion.content;
    $("modalEyebrow").textContent = "Update your post";
    $("modalTitle").textContent = "Edit discussion";
    $("saveButton").textContent = "Save changes";
    $("discussionModal").classList.add("show");
}

function closeDiscussionModal() {
    $("discussionModal").classList.remove("show");
    $("discussionOrganization").disabled = false;
}

async function saveDiscussion(event) {
    event.preventDefault();

    const discussionId = $("discussionId").value;
    const payload = {
        org_id: Number($("discussionOrganization").value),
        title: $("discussionTitle").value.trim(),
        content: $("discussionContent").value.trim(),
        user_id: currentUserId
    };

    const url = discussionId ? `/api/discussions/${discussionId}` : "/api/discussions";
    const method = discussionId ? "PUT" : "POST";

    try {
        const response = await fetch(url, {
            method,
            headers: {"Content-Type": "application/json"},
            body: JSON.stringify(payload)
        });
        const data = await response.json();
        if (!response.ok) {
            throw new Error(data.message || "Unable to save discussion.");
        }

        closeDiscussionModal();
        selectedDiscussionId = data.discussion.discussion_id;
        toast(data.message || "Discussion saved.");
        await loadDiscussions();
        renderDiscussionDetail(data.discussion);
    } catch (error) {
        toast(error.message || "Unable to save discussion.");
    }
}

async function deleteDiscussion(discussionId) {
    if (!confirm("Delete this discussion?")) {
        return;
    }

    try {
        const response = await fetch(`/api/discussions/${discussionId}?user_id=${currentUserId}`, {
            method: "DELETE"
        });
        const data = await response.json();
        if (!response.ok) {
            throw new Error(data.message || "Unable to delete discussion.");
        }

        selectedDiscussionId = null;
        toast(data.message || "Discussion deleted.");
        await loadDiscussions();
    } catch (error) {
        toast(error.message || "Unable to delete discussion.");
    }
}

async function logoutUser() {
    localStorage.removeItem("user");
    try {
        await fetch("http://127.0.0.1:5001/logout", {
            method: "GET",
            credentials: "include"
        });
    } catch (error) {
        console.error("Logout failed:", error);
    }
    window.location.href = "http://127.0.0.1:5001/";
}

function scheduleSearch() {
    clearTimeout(searchTimer);
    searchTimer = setTimeout(loadDiscussions, 250);
}

document.addEventListener("DOMContentLoaded", async () => {
    const loggedIn = await requireLoggedInUser();
    if (!loggedIn) {
        return;
    }

    await loadOrganizations();
    await loadDiscussions();

    $("searchInput").addEventListener("input", scheduleSearch);
    $("organizationFilter").addEventListener("change", loadDiscussions);
    $("discussionForm").addEventListener("submit", saveDiscussion);
    $("discussionModal").addEventListener("click", (event) => {
        if (event.target === $("discussionModal")) {
            closeDiscussionModal();
        }
    });
});

File: Backend/discussion Module/discussion.py
import sys
from datetime import date, datetime
from pathlib import Path

from flask import Flask, jsonify, render_template, request

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from Database_Module import connection


app = Flask(__name__, template_folder=".", static_folder=".", static_url_path="/static")


def json_response(status, message=None, **payload):
    response = {"status": status}
    if message:
        response["message"] = message
    response.update(payload)
    return jsonify(response)


def format_date(value):
    if isinstance(value, datetime):
        return value.strftime("%d %b %Y, %I:%M %p")
    if isinstance(value, date):
        return value.strftime("%d %b %Y")
    return str(value) if value else ""


def current_user_id(data=None):
    data = data or {}
    raw_user_id = request.args.get("user_id") or data.get("user_id") or request.form.get("user_id")
    try:
        return max(1, int(raw_user_id))
    except (TypeError, ValueError):
        return None


def discussion_payload(row, active_user_id):
    return {
        "discussion_id": row["discussion_id"],
        "org_id": row["org_id"],
        "organization_name": row["organization_name"],
        "organization_category": row.get("organization_category") or "Wellness",
        "user_id": row["user_id"],
        "author": row["author"],
        "title": row["title"],
        "content": row["content"],
        "created_at": format_date(row["created_at"]),
        "is_owner": bool(active_user_id and int(row["user_id"]) == int(active_user_id)),
    }


def fetch_discussion(cursor, discussion_id):
    cursor.execute(
        """
        SELECT
            d.discussion_id,
            d.org_id,
            d.user_id,
            d.title,
            d.content,
            d.created_at,
            u.full_name AS author,
            o.name AS organization_name,
            o.category AS organization_category
        FROM discussions d
        JOIN users u ON u.user_id = d.user_id
        JOIN organizations o ON o.org_id = d.org_id
        WHERE d.discussion_id = %s
        """,
        (discussion_id,)
    )
    return cursor.fetchone()


@app.route("/", methods=["GET"])
def discussion_page():
    return render_template("discussion.html")


@app.route("/api/organizations", methods=["GET"])
def list_organizations():
    conn = None
    cursor = None
    try:
        conn = connection()
        if conn is None:
            return json_response("error", "Database connection failed."), 500

        cursor = conn.cursor(dictionary=True)
        cursor.execute(
            """
            SELECT org_id, name, category, privacy
            FROM organizations
            ORDER BY name ASC
            """
        )
        return json_response("success", organizations=cursor.fetchall())
    except Exception as error:
        return json_response("error", "Unable to load organizations.", error=str(error)), 500
    finally:
        if cursor:
            cursor.close()
        if conn:
            conn.close()


@app.route("/api/discussions", methods=["GET"])
def list_discussions():
    conn = None
    cursor = None
    try:
        conn = connection()
        if conn is None:
            return json_response("error", "Database connection failed."), 500

        active_user_id = current_user_id()
        search = request.args.get("search", "").strip()
        org_id = request.args.get("org_id", "").strip()
        where_parts = []
        params = []

        if search:
            pattern = f"%{search}%"
            where_parts.append("(d.title LIKE %s OR d.content LIKE %s OR u.full_name LIKE %s OR o.name LIKE %s)")
            params.extend([pattern, pattern, pattern, pattern])

        if org_id:
            try:
                where_parts.append("d.org_id = %s")
                params.append(int(org_id))
            except ValueError:
                return json_response("error", "Invalid organization selected."), 400

        where_clause = f"WHERE {' AND '.join(where_parts)}" if where_parts else ""
        cursor = conn.cursor(dictionary=True)
        cursor.execute(
            f"""
            SELECT
                d.discussion_id,
                d.org_id,
                d.user_id,
                d.title,
                d.content,
                d.created_at,
                u.full_name AS author,
                o.name AS organization_name,
                o.category AS organization_category
            FROM discussions d
            JOIN users u ON u.user_id = d.user_id
            JOIN organizations o ON o.org_id = d.org_id
            {where_clause}
            ORDER BY d.created_at DESC
            """,
            tuple(params)
        )
        discussions = [discussion_payload(row, active_user_id) for row in cursor.fetchall()]
        return json_response("success", discussions=discussions)
    except Exception as error:
        return json_response("error", "Unable to load discussions.", error=str(error)), 500
    finally:
        if cursor:
            cursor.close()
        if conn:
            conn.close()


@app.route("/api/discussions/<int:discussion_id>", methods=["GET"])
def get_discussion(discussion_id):
    conn = None
    cursor = None
    try:
        conn = connection()
        if conn is None:
            return json_response("error", "Database connection failed."), 500

        active_user_id = current_user_id()
        cursor = conn.cursor(dictionary=True)
        discussion = fetch_discussion(cursor, discussion_id)
        if not discussion:
            return json_response("error", "Discussion not found."), 404

        return json_response("success", discussion=discussion_payload(discussion, active_user_id))
    except Exception as error:
        return json_response("error", "Unable to load discussion.", error=str(error)), 500
    finally:
        if cursor:
            cursor.close()
        if conn:
            conn.close()


@app.route("/api/discussions", methods=["POST"])
def create_discussion():
    data = request.get_json(silent=True) or {}
    user_id = current_user_id(data)
    title = str(data.get("title", "")).strip()
    content = str(data.get("content", "")).strip()

    try:
        org_id = int(data.get("org_id", 0))
    except (TypeError, ValueError):
        return json_response("error", "Please choose a valid organization."), 400

    if not user_id:
        return json_response("error", "Please login before creating a discussion."), 401
    if len(title) < 5:
        return json_response("error", "Discussion title must be at least 5 characters."), 400
    if len(title) > 150:
        return json_response("error", "Discussion title must be 150 characters or less."), 400
    if len(content) < 10:
        return json_response("error", "Discussion content must be at least 10 characters."), 400

    conn = None
    cursor = None
    try:
        conn = connection()
        if conn is None:
            return json_response("error", "Database connection failed."), 500

        cursor = conn.cursor(dictionary=True)
        cursor.execute("SELECT user_id FROM users WHERE user_id = %s", (user_id,))
        if not cursor.fetchone():
            return json_response("error", "User not found. Please login first."), 404

        cursor.execute("SELECT org_id FROM organizations WHERE org_id = %s", (org_id,))
        if not cursor.fetchone():
            return json_response("error", "Organization not found."), 404

        cursor.execute(
            """
            INSERT INTO discussions (org_id, user_id, title, content)
            VALUES (%s, %s, %s, %s)
            """,
            (org_id, user_id, title, content)
        )
        discussion_id = cursor.lastrowid
        conn.commit()

        discussion = fetch_discussion(cursor, discussion_id)
        return json_response(
            "success",
            "Discussion created successfully.",
            discussion=discussion_payload(discussion, user_id)
        ), 201
    except Exception as error:
        if conn:
            conn.rollback()
        return json_response("error", "Unable to create discussion.", error=str(error)), 500
    finally:
        if cursor:
            cursor.close()
        if conn:
            conn.close()


@app.route("/api/discussions/<int:discussion_id>", methods=["PUT"])
def update_discussion(discussion_id):
    data = request.get_json(silent=True) or {}
    user_id = current_user_id(data)
    title = str(data.get("title", "")).strip()
    content = str(data.get("content", "")).strip()

    if not user_id:
        return json_response("error", "Please login before editing this discussion."), 401
    if len(title) < 5:
        return json_response("error", "Discussion title must be at least 5 characters."), 400
    if len(title) > 150:
        return json_response("error", "Discussion title must be 150 characters or less."), 400
    if len(content) < 10:
        return json_response("error", "Discussion content must be at least 10 characters."), 400

    conn = None
    cursor = None
    try:
        conn = connection()
        if conn is None:
            return json_response("error", "Database connection failed."), 500

        cursor = conn.cursor(dictionary=True)
        cursor.execute("SELECT user_id FROM discussions WHERE discussion_id = %s", (discussion_id,))
        discussion = cursor.fetchone()
        if not discussion:
            return json_response("error", "Discussion not found."), 404
        if int(discussion["user_id"]) != int(user_id):
            return json_response("error", "Only the author can edit this discussion."), 403

        cursor.execute(
            """
            UPDATE discussions
            SET title = %s, content = %s
            WHERE discussion_id = %s
            """,
            (title, content, discussion_id)
        )
        conn.commit()

        updated = fetch_discussion(cursor, discussion_id)
        return json_response(
            "success",
            "Discussion updated successfully.",
            discussion=discussion_payload(updated, user_id)
        )
    except Exception as error:
        if conn:
            conn.rollback()
        return json_response("error", "Unable to update discussion.", error=str(error)), 500
    finally:
        if cursor:
            cursor.close()
        if conn:
            conn.close()


@app.route("/api/discussions/<int:discussion_id>", methods=["DELETE"])
def delete_discussion(discussion_id):
    user_id = current_user_id()
    if not user_id:
        return json_response("error", "Please login before deleting this discussion."), 401

    conn = None
    cursor = None
    try:
        conn = connection()
        if conn is None:
            return json_response("error", "Database connection failed."), 500

        cursor = conn.cursor(dictionary=True)
        cursor.execute("SELECT user_id FROM discussions WHERE discussion_id = %s", (discussion_id,))
        discussion = cursor.fetchone()
        if not discussion:
            return json_response("error", "Discussion not found."), 404
        if int(discussion["user_id"]) != int(user_id):
            return json_response("error", "Only the author can delete this discussion."), 403

        cursor.execute("DELETE FROM discussions WHERE discussion_id = %s", (discussion_id,))
        conn.commit()
        return json_response("success", "Discussion deleted successfully.")
    except Exception as error:
        if conn:
            conn.rollback()
        return json_response("error", "Unable to delete discussion.", error=str(error)), 500
    finally:
        if cursor:
            cursor.close()
        if conn:
            conn.close()


if __name__ == "__main__":
    print("=" * 60)
    print("          AI WELLNESS DISCUSSIONS")
    print("=" * 60)
    print("Server:")
    print("http://127.0.0.1:5003")
    print()
    print("Discussions API:")
    print("GET/POST http://127.0.0.1:5003/api/discussions")
    print("=" * 60)
    app.run(debug=True, port=5003, use_reloader=False)
